sw/device/lib/dif/dif_usbdev.h

```diff
@@ -0,0 +1,759 @@
+ * - `kDifUsbdevInPacketWriteBytesResultBufferFull` if the requested write
+ * exceeds the device buffer size,
+ * - `kDifUsbdevInPacketWriteBytesResultOK` otherwise.
+ */
+dif_usbdev_tx_packet_write_bytes_result_t dif_usbdev_tx_packet_write_bytes(
+    dif_usbdev_t *usbdev, uint32_t endpoint, uint8_t *src, uint32_t src_len,
+    uint32_t *bytes_written);
+
+/**
+ * Return codes for `dif_usbdev_tx_packet_send`.
+ */
+typedef enum dif_usbdev_tx_packet_send_result {
+  kDifUsbdevInPacketSendResultOK,
+  kDifUsbdevInPacketSendResultInvalidArgument,
+  kDifUsbdevInPacketSendResultGeneralError,
+  kDifUsbdevInPacketSendResultNotWriting,
+} dif_usbdev_tx_packet_send_result_t;
+
+/**
+ * Mark the packet buffered for the given endpoint as ready for transmission.
+ *
+ * See also: `dif_usbdev_tx_packet_write_bytes`.
+ *
+ * @param usbdev A USB device.
+ * @param endpoint An endpoint.
+ * @return
+ * - `kDifUsbdevInPacketSendResultInvalidArgument` if `usbdev` is `NULL`,
+ * - `kDifUsbdevInPacketSendResultGeneralError` if `endpoint` is invalid,
+ * - `kDifUsbdevInPacketSendResultNotWriting` if the `endpoint` does not have a
+ * packet to be sent,
+ * - `kDifUsbdevInPacketSendResultOK` otherwise.
+ */
+dif_usbdev_tx_packet_send_result_t dif_usbdev_tx_packet_send(
+    dif_usbdev_t *usbdev, uint32_t endpoint);
+
+/**
+ * Status of an outgoing packet.
+ */
+typedef enum dif_usbdev_tx_packet_status {
+  // There is no packet for the given endpoint
+  kDifUsbdevTxPacketStatusNoPacket,
+  // There is an active write buffer operation and the packet is not marked as
+  // ready for transmission yet
+  kDifUsbdevTxPacketStatusStillWriting,
+  // Packet is pending transmission
+  kDifUsbdevTxPacketStatusPending,
+  // Packet was sent successfully
+  kDifUsbdevTxPacketStatusSent,
+  // Transmission was canceled due to an incoming SETUP packet
+  kDifUsbdevTxPacketStatusCancelled,
+} dif_usbdev_tx_packet_status_t;
+
+/**
+ * Return codes for `dif_usbdev_tx_packet_get_status`.
+ */
+typedef enum dif_usbdev_tx_packet_get_status_result {
+  kDifUsbdevTxPacketGetStatusResultOK,
+  kDifUsbdevTxPacketGetStatusResultInvalidArgument,
+  kDifUsbdevTxPacketGetStatusResultGeneralError,
+} dif_usbdev_tx_packet_get_status_result_t;
+
+/**
+ * Get the status of a packet that has been queued to be sent from an endpoint.
+ *
+ * @param usbdev A USB device.
+ * @param endpoint An endpoint.
+ * @param status Status of the packet.
+ * @return
+ * - `kDifUsbdevTxPacketGetStatusResultInvalidArgument` if `usbdev` or
+ * `status` is `NULL`,
+ * - `kDifUsbdevTxPacketGetStatusResultGeneralError` if `endpoint` is invalid,
+ * - `kDifUsbdevTxPacketGetStatusResultOK` otherwise.
+ */
+dif_usbdev_tx_packet_get_status_result_t dif_usbdev_tx_packet_get_status(
+    dif_usbdev_t *usbdev, uint32_t endpoint,
+    dif_usbdev_tx_packet_status_t *status);
+
+/**
+ * Set the address of a USB device.
+ *
+ * @param usbdev A USB device.
+ * @param addr New address. Only the last 7 bits are significant.
```

Comment:
> Only the last 7 bits are significant

Why not `uint8_t` then?

Reply:
I was on the fence about this because of our word size and other changes that needed to be made to be consistent. Done. Also made changes to fifo depth, endpoint, frame index, etc., but left IRQ-related code as is to be consistent with other DIF libraries.